src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+<?php
+
+/**
+* @file
+* Contains \Drupal\commerce_heartland\Controller\Heartland.php
+*/
+
+namespace Drupal\commerce_heartland\Plugin\Commerce\PaymentGateway;
+
+use Drupal\commerce_payment\CreditCard;
+use Drupal\commerce_payment\Entity\PaymentInterface;
+use Drupal\commerce_payment\Entity\PaymentMethodInterface;
+use Drupal\commerce_payment\Exception\HardDeclineException;
+use Drupal\commerce_payment\PaymentMethodTypeManager;
+use Drupal\commerce_payment\PaymentTypeManager;
+use Drupal\commerce_payment\Plugin\Commerce\PaymentGateway\OnsitePaymentGatewayBase;
+use Drupal\commerce_payment\Exception\PaymentGatewayException;
+use Drupal\commerce_price\Price;
+use Drupal\Component\Datetime\TimeInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Form\FormStateInterface;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+
+use GlobalPayments\Api\PaymentMethods\CreditCardData;
+use GlobalPayments\Api\Services\CreditService;
+use GlobalPayments\Api\ServicesConfig;
+use GlobalPayments\Api\ServicesContainer;
+use GlobalPayments\Api\Entities\Exceptions\GatewayException;
+use GlobalPayments\Api\Entities\Transaction;
+use GlobalPayments\Api\Entities\Enums\PaymentMethodType;
+use GlobalPayments\Api\Services\ReportingService;
+
+/**
+ * Provides the On-site payment gateway.
+ *
+ * @CommercePaymentGateway(
+ *   id = "commerce_heartland",
+ *   label = "Heartland",
+ *   display_label = "Token-based payment gateway from Heartland",
+ *   forms = {
+ *     "add-payment-method" = "Drupal\commerce_heartland\PluginForm\Onsite\PaymentMethodAddForm",
+ *   },
+ *   payment_method_types = {"credit_card"},
+ *   credit_card_types = {
+ *     "amex", "dinersclub", "discover", "jcb", "maestro", "mastercard", "visa",
+ *   },
+ *   js_library = "commerce_heartland/form"
+ * )
+ */
+
+
+class Heartland extends OnsitePaymentGatewayBase implements OnsiteInterface
+{
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityTypeManagerInterface $entity_type_manager, PaymentTypeManager $payment_type_manager, PaymentMethodTypeManager $payment_method_type_manager, TimeInterface $time)
+    {
+        parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_type_manager, $payment_type_manager, $payment_method_type_manager, $time);
+  
+        // Heartland SDK - Authentication
+        // https://developer.heartlandpaymentsystems.com/Documentation/authentication/#authentication
+        $config = new ServicesConfig();
+        $config->secretApiKey = $this->configuration['secret_key'];
+        $config->serviceUrl = "https://cert.api2.heartlandportico.com";
```

Comment:
This needs to be based on the secret API key's environment:

- `skapi_cert_*` -> `https://cert.api2-c.heartlandportico.com`
- `skapi_prod_*` -> `https://api2-c.heartlandportico.com`